### PR TITLE
Revert with reason

### PR DESCRIPTION
**Pull request by chriseth**
Part of issue https://github.com/ethereum/solidity/issues/1686

Depends on https://github.com/ethereum/solidity/pull/3373

I went with encoding the reason string as a function call to `Erorr(string)`, including the function selector.

Still to do:

 - [x] the same for ``require()``
 - [x] documentation for ``require()``
 - [x] forwarding of the reason string by the caller together with bubbling up the error
 - [x] tests for forwarding reason string for weird cases like create and transfer.

The reason we might not want to forward the reason string is that it allows a called attacker to consume more gas in the caller (because the caller needs to copy the return data), but I think this is an acceptable "risk" given the benefit that the error string shows up in the transaction receipt.

### CODE REVIEW

**Comment by SilentCicero** [comment]
Looking forward to this ;)

**Comment by GriffGreen** [comment]
Me toooooooo!!!! :-D

**Comment by chriseth** [comment]
This is ready for review.

**Comment by onbjerg** [comment]
❤️ ❤️ ❤️ ❤️ This is the best!

**Comment by izqui** [comment]
Is the string returned regularly, right? In terms of forwarding the return data, would this work https://github.com/aragon/aragon-core/blob/dev/contracts/common/DelegateProxy.sol#L9

**Comment by chriseth** [comment]
Rebased.

**Comment by AnthonyAkentiev** [comment]
Great improvement!

**Comment by tcoulter** [comment]
Looking forward to implementing this in Truffle. Nice job @chriseth!

**Comment by naddison36** [comment]
Sad to see this bumped to 0.4.22. I was getting ready to change my contracts to take advantage of this new feature with the 0.4.21 release

**Comment by chriseth** [comment]
Rebased and moved changelog entry.

**Comment by chriseth** [comment]
Will rebase again.

**Comment by axic** [comment]
Rebased.

File: test/libsolidity/Assembly.cpp [hunk 1]
```diff
@@ -158,13 +158,24 @@ BOOST_AUTO_TEST_CASE(location_test)
 	}
 	)";
 	shared_ptr<string const> n = make_shared<string>("");
+	shared_ptr<string const> codegen = make_shared<string>("--CODEGEN--:8-17");
```
Comment (axic) [inline test/libsolidity/Assembly.cpp]: This variable is not used.

File: docs/control-structures.rst [hunk 2]
```diff
@@ -469,13 +470,16 @@ of an exception instead of "bubbling up".
 Catching exceptions is not yet possible.
 
 In the following example, you can see how ``require`` can be used to easily check conditions on inputs
-and how ``assert`` can be used for internal error checking::
+and how ``assert`` can be used for internal error checking. Note that you can optionally provide
+a message string for require, but not for assert.
```
Comment (axic) [inline docs/control-structures.rst]: Need to "code-escape" require and assert.

File: docs/control-structures.rst [hunk 3]
```diff
@@ -513,3 +517,29 @@ the EVM to revert all changes made to the state. The reason for reverting is tha
...
+            // Alternative way to do it:
+            require(
+                amount <= msg.value / 2 ether,
+                "Not enough Ether provided."
+            );
+            // Perform the purchase.
+        }
+    }
+
+The provided string will be abi-encoded together with a uint value that will always be zero.
```
Comment (axic) [inline docs/control-structures.rst]: ABI-encoded and should make that into a link to the ABI spec

File: libsolidity/ast/AST.cpp [hunk 4]
```diff
@@ -338,6 +338,7 @@ shared_ptr<FunctionType> FunctionDefinition::functionType(bool _internal) const
 
 TypePointer FunctionDefinition::type() const
 {
+	solAssert(visibility() != Declaration::Visibility::External, "");
```
Comment (axic) [inline libsolidity/ast/AST.cpp:341]: Why is this needed?

Comment (chriseth) [reply libsolidity/ast/AST.cpp:341]: That's a good question. I don't remember, but it does not seem to fail...

Comment (chriseth) [reply libsolidity/ast/AST.cpp:341]: Ah, the `FunctionType` constructor has a second argument that tells whether we want to have the type of the internal aspect or the type of the external aspect. It defaults to "internal".

**Comment by axic** [comment]
How about changing this to be semi-compatible with EIP-838: https://github.com/ethereum/EIPs/issues/838#issuecomment-379238333

**Comment by axic** [comment]
To clarify: lets just use `Error(string)` and replace the current encoding, but do not change any other part of the PR (e.g. do not add the internal error messages yet).

File: docs/control-structures.rst [hunk 5]
```diff
@@ -515,3 +519,29 @@ the EVM to revert all changes made to the state. The reason for reverting is tha
...
```
Comment (chriseth) [inline docs/control-structures.rst]: Change this.

**Comment by chriseth** [comment]
@axic updated.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 6]
```diff
@@ -900,16 +922,42 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
...
 			auto success = m_context.appendConditionalJump();
 			if (function.kind() == FunctionType::Kind::Assert)
 				// condition was not met, flag an error
 				m_context.appendInvalid();
+			else if (arguments.size() > 1)
+			{
+				m_context << u256(0);
+				utils().moveIntoStack(arguments.at(1)->annotation().type->sizeOnStack(), 1);
+				utils().fetchFreeMemoryPointer();
+				utils().abiEncode(
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp]: I think this wasn't updated to use the new encoding?

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Oh wow, sorry for not having tests for that...

File: libsolidity/analysis/NameAndTypeResolver.cpp [hunk 7]
```diff
@@ -47,7 +47,9 @@ NameAndTypeResolver::NameAndTypeResolver(
 	if (!m_scopes[nullptr])
 		m_scopes[nullptr].reset(new DeclarationContainer());
 	for (Declaration const* declaration: _globals)
-		m_scopes[nullptr]->registerDeclaration(*declaration);
+	{
```
Comment (ekpyron) [inline libsolidity/analysis/NameAndTypeResolver.cpp:50]: Superfluous brackets.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 6 above)
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Ah right, there are tests, I just did not update them :)

**Comment by chriseth** [comment]
Updated.

File: libsolidity/codegen/CompilerUtils.cpp [hunk 8]
```diff
@@ -78,6 +78,19 @@ void CompilerUtils::toSizeAfterFreeMemoryPointer()
 	m_context << Instruction::SWAP1;
 }
 
+void CompilerUtils::revertWithStringData(Type const& _argumentType)
+{
```
Comment (ekpyron) [inline libsolidity/codegen/CompilerUtils.cpp:82]: Maybe an assertion about the type of ``_argumentType``?

**Review by ekpyron — DISMISSED** [review]
I added some optional minor suggestions, but in general this looks fine to me!

File: libsolidity/analysis/NameAndTypeResolver.cpp (hunk 7 above)
Comment (chriseth) [reply libsolidity/analysis/NameAndTypeResolver.cpp:50]: Some compilers, most notably MSVC have problems with macros if they are the single statement in a loop body. Because of that, I always put braces around macros in such cases.

File: libsolidity/codegen/CompilerUtils.cpp (hunk 8 above)
Comment (chriseth) [reply libsolidity/codegen/CompilerUtils.cpp:82]: `abiEncode` will complain if the conversion is not possible, but probably still a good idea.

**Comment by chriseth** [comment]
@ekpyron alex was mainly concerned about the general concept of how the error string is encoded. What do you think?

File: libsolidity/analysis/NameAndTypeResolver.cpp (hunk 7 above)
Comment (ekpyron) [reply libsolidity/analysis/NameAndTypeResolver.cpp:50]: Ah, OK, that makes sense.

**Comment by ekpyron** [comment]
@chriseth I think ABI-encoding as a function call to some error function makes sense. I was wondering about the ``Error(string)`` signature and whether it should be a different name, resp. different for ``revert`` and ``require``, but I think using the same makes sense and the name ``Error`` makes sense as well.

Were there any specific concerns? What would be the alternative? Just returning the encoded error string directly?

**Comment by ekpyron** [comment]
Just reading through https://github.com/ethereum/EIPs/issues/838#issuecomment-379238333 again.

**Comment by ekpyron** [comment]
If I understand https://github.com/ethereum/EIPs/issues/838#issuecomment-379238333 and the connection to this PR correctly, it seems to me that the current concept makes a lot of sense.

**Comment by ekpyron** [comment]
If the concern is about whether to use something like ``Error(cstring)`` or ``Error(bytes)`` instead of ``Error(string)`` to save space in the encoding, I think the fact, that ``Error(string)`` is more natural and it is the signature I would usually expect for a function accepting a string, is more important than saving the space. Especially since in the future the concept can be extended by specific errors with different signatures that may be more space-efficient, following https://github.com/ethereum/EIPs/issues/838#issuecomment-379238333.

**Comment by chriseth** [comment]
Failure in appveyor is just bytecode upload and thus can be ignored.

**Comment by chriseth** [comment]
I think prefixing a specific 4-byte sequence allows us to extend the encoding later, while at the same time it does not reduce usability since it is fixed for now. Also, it is better than just prefixing a certain number of zeros, since it already suggests away how to extend it to encode different types.

I would opt to merge it :)

**Review by ekpyron — DISMISSED** [review]

File: docs/control-structures.rst [hunk 9]
```diff
@@ -455,8 +455,9 @@ The ``require`` function should be used to ensure valid conditions, such as inpu
 If used properly, analysis tools can evaluate your contract to identify the conditions and function calls which will reach a failing ``assert``. Properly functioning code should never reach a failing assert statement; if this happens there is a bug in your contract which you should fix.
 
 There are two other ways to trigger exceptions: The ``revert`` function can be used to flag an error and
-revert the current call. In the future it might be possible to also include details about the error
-in a call to ``revert``. The ``throw`` keyword can also be used as an alternative to ``revert()``.
+revert the current call. It is possible to provide a string message containing details about the error
+that will be passed back to the caller.
+The ``throw`` keyword can also be used as an alternative to ``revert()``, but is deprecated.
```
Comment (axic) [inline docs/control-structures.rst]: ... and it doesn't support an error string.

Comment (chriseth) [reply docs/control-structures.rst]: Yes, alternative to `rever()`, not to `revert("string")` :)
Do you want me to clarify that or not?

File: test/libsolidity/Assembly.cpp [hunk 10]
```diff
@@ -160,11 +160,21 @@ BOOST_AUTO_TEST_CASE(location_test)
 	shared_ptr<string const> n = make_shared<string>("");
```
Comment (axic) [inline test/libsolidity/Assembly.cpp]: This is still dead code.

File: docs/control-structures.rst (hunk 9 above)
Comment (axic) [reply docs/control-structures.rst]: I think it may be cleaner that way. I didn't spot it before you mentioned, of course it is clear after you did.

**Review by axic — CHANGES_REQUESTED** [review]
Needs test cases for built in revert bubbling:
- `transfer`
- contract creation
- regular foreign contract calls

i.e.
```
contract C {
  function f() {
    revert("failed here :(");
  }
}
contract D {
  D d = new D();
  function f() {
    return d.f();
  }
}
```

**Comment by chriseth** [comment]
The test case `bubble_up_error_messages` does it for a regular external function call, `bubble_up_error_messages_through_transfer` for transfer and `bubble_up_error_messages_through_create` for create....

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
That's weird, I have seen revert_with_cause and require_with_message, but not the bubble up ones. Anyway, they look correct.

**Comment by leanthebean** [comment]
Oh wow, this is so great. Thank you!

**Comment by vsdigitall** [comment]
```
contract Example {
  function test (uint i) {
    require(i == 1, "ERROR_TEXT");
  }
}
```

How to read require/revert error string from failed transaction log? (ERROR_TEXT in example above)

**Comment by andreafspeziale** [comment]
I know that maybe is not the right place but I've not found any references or examples so +1 to @vsdigitall. I think that the reason string is in the output as hex in the failed transaction. But I've been not able to fully test this thing.

**Comment by gluk64** [comment]
Here is a bash script to fetch revert reason:

https://ethereum.stackexchange.com/a/70391/3558
